testutil/validatormock/synccontributionflow_test.go

```diff
@@ -0,0 +1,315 @@
+	"testing"
+
+	eth2v1 "github.com/attestantio/go-eth2-client/api/v1"
+	"github.com/attestantio/go-eth2-client/mock"
+	"github.com/attestantio/go-eth2-client/spec/altair"
+	eth2p0 "github.com/attestantio/go-eth2-client/spec/phase0"
+
+	"github.com/obolnetwork/charon/eth2util/signing"
+)
+
+// PseudoSyncCommContributionFlow is an example of how the Sync Committee Contribution
+// flow works for the v1 and DVT v2 beacon api.
+func PseudoSyncCommContributionFlow(t *testing.T, supportDVT bool) {
+	// See spec: https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/validator.md#sync-committees
+
+	ctx := context.Background()
+	eth2Cl, _ := mock.New(ctx) // Error handling omitted for brevity.
+
+	// Given a set of validators with private keys
+	var valIdxs []eth2p0.ValidatorIndex
+	var signFunc SignFunc
+
+	// For some epoch, either the current or the next epoch.
+	var epoch eth2p0.Epoch
+
+	// Calculate current sync committee period start and end.
+	spec, _ := eth2Cl.Spec(ctx)
+	periodInt, _ := spec["EPOCHS_PER_SYNC_COMMITTEE_PERIOD"].(int)
+
+	period := eth2p0.Epoch(periodInt)
+
+	startEpoch := (epoch / period) * period
+	endEpoch := startEpoch + period
+
+	// Get the sync committee duties for this epoch.
+
+	// One option is to fetch the sync committee duties for a subset of validators
+	duties, _ := eth2Cl.SyncCommitteeDuties(ctx, epoch, valIdxs)
+	for _, duty := range duties {
+		// Note SyncCommitteeDuty contains a public key which charon needs to intercept/swap.
+		// Note that each validator could have multiple indeces in the sync committee.
+		// Note that validators not in the committee is omitted from the response.
+		t.Logf("SyncComm Duty VIdx=%v, ValSyncCommIdx=%v",
+			duty.ValidatorIndex, duty.ValidatorSyncCommitteeIndices)
+	}
+
+	// Another option is to fetch the whole sync committee duty itself.
+	// Note that the ValidatorAggregates field is probably the subnets...?
+	syncCommittee, _ := eth2Cl.SyncCommitteeAtEpoch(ctx, "head", epoch)
+
+	t.Logf("SyncCommittee len(Validators)=%v, len(ValidatorAggregates)=%v",
+		len(syncCommittee.Validators), len(syncCommittee.ValidatorAggregates))
+	// Lookup validators in the returned lists.
+
+	// At the start of each epoch in the period, notify the BN to subscribe to the subnet.
+	// Doing this repeatedly each epoch in case the BN restarts.
+
+	// Create and then submit subscriptions
+	var subs []*eth2v1.SyncCommitteeSubscription
+	for _, duty := range duties {
+		subs = append(subs, &eth2v1.SyncCommitteeSubscription{
+			ValidatorIndex:       duty.ValidatorIndex,
+			SyncCommitteeIndices: duty.ValidatorSyncCommitteeIndices,
+			UntilEpoch:           endEpoch,
+		})
+	}
+	_ = eth2Cl.SubmitSyncCommitteeSubscriptions(ctx, subs)
+
+	// At 1/3 into each slot in sync committee period, submit a sync committee message.
+	// Get the current head block root (note there are probably better ways to do this).
+	var slot eth2p0.Slot // Note the spec mentioned something about previous slot (slot-1).
```

Comment:
From the specs:
If a validator is in the current sync committee (i.e. `is_assigned_to_sync_committee()` above returns `True`), then for every `slot` in the current sync committee period, the validator should prepare a `SyncCommitteeMessage` for the previous slot (`slot - 1`) according to the logic in `get_sync_committee_message` as soon as they have determined the head block of `slot - 1`. This means that when assigned to `slot` a `SyncCommitteeMessage` is prepared and broadcast in `slot-1 ` instead of `slot`.

Reply:
So, as i understand:
If a validator is assigned to a sync committee from epochs [256-511], it must produce and broadcast `SyncCommitteeMessage` in the last slot of epoch 255, ie, slot `8191`

Reply:
agreed